src/Features/CSharp/Portable/SignatureHelp/AbstractCSharpSignatureHelpProvider.cs

```diff
@@ -68,5 +96,287 @@ protected static SymbolDisplayPart NewLine()
         protected IList<TaggedText> GetAwaitableUsage(IMethodSymbol method, SemanticModel semanticModel, int position)
 #pragma warning restore CA1822 // Mark members as static
             => SpecializedCollections.EmptyList<TaggedText>();
+
+        /// <summary>
+        /// If the symbol could not be bound, we could be dealing with a partial invocation, we'll try to find a possible overload.
+        /// </summary>
+        protected static (IMethodSymbol? symbol, int parameterIndex) GuessCurrentSymbolAndParameter(
+            SeparatedSyntaxList<ArgumentSyntax> arguments, ImmutableArray<IMethodSymbol> methodGroup, int position,
+            SemanticModel semanticModel, ISemanticFactsService semanticFactsService)
+        {
+            if (arguments.Count != 0)
+            {
+                foreach (var method in methodGroup)
+                {
+                    if (FindParameterIndexIfCompatibleMethod(arguments, method, position, semanticModel, semanticFactsService, out var parameterIndex))
+                    {
+                        return (method, parameterIndex);
+                    }
+                }
+            }
+
+            // Note: Providing no recommendation if no arguments allows the model to keep the last implicit choice
+            return (null, -1);
+        }
+
+        /// <summary>
+        /// Simulates overload resolution with the arguments provided so far and determines if you might be calling this overload.
+        /// Returns true if an overload is acceptable. In that case, we output the parameter should be highlighted given the cursor's
+        /// position in the partial invocation.
+        /// </summary>
+        protected static bool FindParameterIndexIfCompatibleMethod(SeparatedSyntaxList<ArgumentSyntax> arguments, IMethodSymbol method, int position,
+            SemanticModel semanticModel, ISemanticFactsService semanticFactsService, out int foundParameterIndex)
+        {
+            // map the arguments to their corresponding parameters
+            var argToParamMap = PrepareArgToParamMap(arguments, method);
```

Comment:
can you not put the map into a `using` so that all codepaths don't need to free it?